bytecomp/bytegen.ml

```diff
@@ -897,19 +897,29 @@ let rec comp_expr env exp sz cont =
           let c = comp_expr env lam sz cont in
           let ev = event Event_pseudo Event_other in
           add_event ev c
-      | Lev_after _ when is_tailcall cont -> (* don't destroy tail call opt *)
-          comp_expr env lam sz cont
       | Lev_after ty ->
-          let info =
+          let preserve_tailcall =
             match lam with
-              Lapply{ap_args = args}  -> Event_return (List.length args)
-            | Lsend(_, _, _, args, _) -> Event_return (List.length args + 1)
-            | Lprim(_,args,_)         -> Event_return (List.length args)
-            | _                       -> Event_other
+            | Lprim((Pidentity | Popaque | Pdirapply | Prevapply), _, _) ->
+               true
+            | Lprim(_, _, _) -> false
+            | _ -> true
           in
-          let ev = event (Event_after ty) info in
-          let cont1 = add_event ev cont in
-          comp_expr env lam sz cont1
+          if not preserve_tailcall && is_tailcall cont then
+            (* don't destroy tail call opt *)
+            comp_expr env lam sz cont
```

Comment:
Sorry for the naive question, but I don't understand the `not preserve_tailcall` test here. My understanding from the `preserve_tailcall` code is that when it returns `true`, then we *should* preserve tail calls by *not* emitting debug information. But the test here suggests that you skip debug information when `not preserve_tailcall` holds, which is the opposite of what I would expect. (Again, an explanation comment could clarify.)

Reply:
Indeed, you are right. This is completely bogus (and made the testsuite fail in several ways, btw). This is a result of a wrong last-minute refactoring. This is now fixed.